File: solutions/php/rail-fence-cipher/1/RailFenceCipher.php
<?php
declare(strict_types=1);

function encode(string $plainMessage, int $rails): string
{
    $len = strlen($plainMessage);
    $fence = array_fill(0, $rails, array_fill(0, $len, ''));
    $rail = 0;
    $dir = 1; // 1 = down, -1 = up

    for ($i = 0; $i < $len; $i++) {
        $fence[$rail][$i] = $plainMessage[$i];
        $rail += $dir;
        if ($rail === 0 || $rail === $rails - 1) {
            $dir = -$dir;
        }
    }

    $result = '';
    for ($r = 0; $r < $rails; $r++) {
        for ($i = 0; $i < $len; $i++) {
            if ($fence[$r][$i] !== '') {
                $result .= $fence[$r][$i];
            }
        }
    }
    return $result;
}

function decode(string $cipherMessage, int $rails): string
{
    $len = strlen($cipherMessage);
    $fence = array_fill(0, $rails, array_fill(0, $len, ''));

    // Mark positions in zigzag pattern
    $rail = 0;
    $dir = 1;
    for ($i = 0; $i < $len; $i++) {
        $fence[$rail][$i] = '*';
        $rail += $dir;
        if ($rail === 0 || $rail === $rails - 1) {
            $dir = -$dir;
        }
    }

    // Fill marked positions with ciphertext characters rail by rail
    $idx = 0;
    for ($r = 0; $r < $rails; $r++) {
        for ($i = 0; $i < $len; $i++) {
            if ($fence[$r][$i] === '*') {
                $fence[$r][$i] = $cipherMessage[$idx++];
            }
        }
    }

    // Read zigzag to get plaintext
    $result = '';
    $rail = 0;
    $dir = 1;
    for ($i = 0; $i < $len; $i++) {
        $result .= $fence[$rail][$i];
        $rail += $dir;
        if ($rail === 0 || $rail === $rails - 1) {
            $dir = -$dir;
        }
    }
    return $result;
}
?>

File: solutions/ruby/rail-fence-cipher/1/rail_fence_cipher.rb
class RailFenceCipher
  def self.encode(message, rails)
    return '' if message.empty?
    return message if rails <= 1

    # Agregamos con_index para usar la posición original como orden secundario
    rail_pattern(message.length, rails)
      .zip(message.chars)
      .each_with_index
      .sort_by { |(rail, _char), index| [rail, index] }
      .map { |(rail, char), index| char }
      .join
  end

  def self.decode(ciphertext, rails)
    return '' if ciphertext.empty?
    return ciphertext if rails <= 1

    positions = (0...ciphertext.length).to_a
    # Hacemos lo mismo aquí: ordenar por riel, y desempatar por el índice original
    decode_order = rail_pattern(ciphertext.length, rails)
      .zip(positions)
      .each_with_index
      .sort_by { |(rail, _pos), index| [rail, index] }
      .map { |(rail, pos), index| pos }

    result = Array.new(ciphertext.length)
    decode_order.zip(ciphertext.chars).each do |original_pos, char|
      result[original_pos] = char
    end

    result.join
  end

  private

  def self.rail_pattern(length, rails)
    pattern = (0...rails).to_a + (1...rails - 1).to_a.reverse
    pattern.cycle.take(length)
  end
end